Mathlib/Topology/SubsetProperties.lean

```diff
@@ -1307,6 +1309,120 @@ theorem IsClosed.exists_minimal_nonempty_closed_subset [CompactSpace α] {S : Se
   rw [← this, compl_compl]
 #align is_closed.exists_minimal_nonempty_closed_subset IsClosed.exists_minimal_nonempty_closed_subset
 
+/-- A subset `s ⊆ α` is called **σ-compact** if it is the union of countably many compact sets. -/
+def IsSigmaCompact (s : Set α) : Prop :=
+  ∃ K : ℕ → Set α, (∀ n, IsCompact (K n)) ∧ ⋃ n, K n = s
+
+/-- The empty set is σ-compact. -/
+@[simp]
+lemma isSigmaCompact_empty : IsSigmaCompact (∅ : Set α) := by
+  use fun _ ↦ ∅
+  simp only [isCompact_empty, forall_const, iUnion_empty, and_self]
+
+/-- Compact sets are σ-compact. -/
+lemma IsCompact.isSigmaCompact {s : Set α} (hs : IsCompact s) : IsSigmaCompact s :=
+  ⟨fun _ => s, fun _ => hs, iUnion_const _⟩
+
+/-- Countable unions of compact sets are σ-compact. -/
+lemma isSigmaCompact_of_countable_compact {S : Set (Set α)} (hc : Set.Countable S)
+    (hcomp : ∀ (s : Set α), s ∈ S → IsCompact s) : IsSigmaCompact (⋃₀ S) := by
+  by_cases S = ∅
+  · simp only [h, sUnion_empty, isSigmaCompact_empty]
+  · -- If S is non-empty, choose a surjection f : ℕ → S, this yields a map ℕ → Set X.
+    obtain ⟨f, hf⟩ := (Set.countable_iff_exists_surjective (nmem_singleton_empty.mp h)).mp hc
+    refine ⟨fun n ↦ f n, fun n ↦ hcomp (f n) (Subtype.mem (f n)), ?_⟩
+    -- I presume this part can be golfed.
+    apply Subset.antisymm
+    · -- Suppose x ∈ ⋃ n, f n. Then x ∈ f i for some i,
+      intro _ hx
+      rw [mem_iUnion] at hx
+      rcases hx with ⟨i, hi⟩
+      -- ... but f i is a set in S, so x ∈ ⋃ S as well.
+      exact ⟨f i, Subtype.mem (f i), hi⟩
+    · -- Suppose x ∈ ⋃ s, then x ∈ s for some s ∈ S.
+      intro x hx
+      rw [mem_sUnion] at hx
+      rcases hx with ⟨s, h, hxs⟩
+      -- Choose n with f n = s (using surjectivity of f).
+      have : ∃ n, f n = s := by
+        obtain ⟨y, hy⟩ := hf ⟨s, h⟩
+        use y
+        simp_all only [Subtype.forall]
+      rcases this with ⟨n, hn⟩
+      exact ⟨f n, mem_range_self n, (by rw [hn]; exact hxs)⟩
+
+/-- Countable unions of σ-compact sets are σ-compact. -/
+lemma isSigmaCompact_of_countable_sigma_compact (S : Set (Set α)) (hc : Set.Countable S)
```

Comment:
Like above, I would name that either `isSigmaCompact_sUnion` or `Set.Countable.isSigmaCompact_sUnion`. I think it's fine to make less things explicit here than in the previous version because it's strictly stronger so it should be the "default" one.

Reply:
Could you also add an `iUnion` version which would take `[Countable \iota]` as a typeclass parameter, as well as a `biUnion` version analogous to [Set.Finite.isCompact_biUnion](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/SubsetProperties.html#Set.Finite.isCompact_biUnion)? You could also try to prove the indexed union version first and then deduce the `sUnion` version, imo it often gives cleaner proofs.

Reply:
Added both. Indeed, proving the iUnion version first was cleaner.

I've cargo-culted the description for "biUnion" in the docstring. (Is there a central place which should have this explanation?)